src/Tmds.DBus/Connection.cs

```diff
@@ -1040,6 +1040,84 @@ internal async Task<IDisposable> WatchSignalAsync(ObjectPath path, string @inter
             }
         }
 
+        /// <summary>
+        /// Watch the bus for a matching signal.
+        /// </summary>
+        /// <param name="rule">
+        /// The match rules.
+        /// </param>
+        /// <param name="handler">
+        /// Callback that is invoked when a matching signal is received.
+        /// </param>
+        /// <param name="onError">
+        /// Callback that is invoked on error.
+        /// </param>
+        public async Task<IDisposable> WatchSignalAsync<TSignalArgs>(SignalMatchRule rule, Action<(ObjectPath @object, TSignalArgs args)> handler, Action<Exception> onError = null)
+        {
+            var synchronizationContext = CaptureSynchronizationContext();
+            var wrappedDisposable = new WrappedDisposable(synchronizationContext);
+            var readValue = ReadMethodFactory.CreateReadMethodDelegate<TSignalArgs>();
+
+            SignalHandler signalHandler = (msg, ex) =>
+            {
+                if (ex != null)
+                {
+                    if (onError == null)
+                    {
+                        return;
+                    }
+                    wrappedDisposable.Call(onError, ex, disposes: true);
+                    return;
+                }
+
+                var reader = new MessageReader(msg, _factory);
+                var value = readValue(reader);
+
+                var args = new List<object>();
```

Comment:
Can we avoid creating this `List` in cases where `MatchArgs` will return true because there are no arguments to check?

Reply:
I guess we could avoid this altogether if we only allow `arg0*`.

Reply:
Yes, this will go away.